codegen/masm/src/codegen/opt/operands/mod.rs

```diff
@@ -0,0 +1,171 @@
+mod context;
+mod solver;
+mod stack;
+mod tactics;
+
+use self::context::SolverContext;
+pub use self::solver::{OperandMovementConstraintSolver, SolverError};
+use self::stack::Stack;
+
+use std::fmt;
+use std::num::NonZeroU8;
+
+use miden_hir as hir;
+
+/// This represents a specific action that should be taken by
+/// the code generator with regard to an operand on the stack.
+///
+/// The output of the optimizer is a sequence of these actions,
+/// the effect of which is to place all of the current instruction's
+/// operands exactly where they need to be, just when they are
+/// needed.
+#[derive(Debug, Copy, Clone)]
+pub enum Action {
+    /// Copy the operand at the given index to the top of the stack
+    Copy(u8),
+    /// Swap the operand at the given index with the one on top of the stack
+    Swap(u8),
+    /// Move the operand at the given index to the top of the stack
+    MoveUp(u8),
+    /// Move the operand at the top of the stack to the given index
+    MoveDown(u8),
+}
+
+/// This is a [miden_hir::Value], but with a modified encoding that lets
+/// us uniquely identify aliases of a value on the operand stack during
+/// analysis.
+///
+/// Aliases of a value are treated as unique values for purposes of operand
+/// stack management, but are associated with multiple copies of a value
+/// on the stack.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Value(u32);
```

Comment:
I had a similar thought, the problem is that the value isn't necessarily aliased, so calling it `AliasedValue` is a bit of a misnomer. `AliasableValue` felt more like a trait name to me, but might work. In the end I settled with `Value` since it is in fact meant to _be_ an `hir::Value`, just with some additional semantics around aliasing. I'm open to suggestions though, naming things is hard :(.

Reply:
Yes, you're right. `AliasedValue` does not have a good fit. I like `AliasableValue`. Although it sounds like a trait, I think it's better than `Value`. I also see the point in keeping it as `Value`. I'm fine either way.

Reply:
I've decided to rename it to `ValueOrAlias`, which is slightly more verbose, but more precise in terms of what the thing actually represents.